pulp_container/app/tasks/sync_stages.py

```diff
@@ -206,31 +188,78 @@ async def get_signature_source(self):
+                        self.signature_dcs.extend(man_sig_dcs)
                     tag_dc.extra_data["tagged_manifest_dc"] = man_dc
-                    await man_dc.resolution()
                     await self.handle_blobs(man_dc, content_data)
-                    tag_dcs.append(tag_dc)
-                    await pb_parsed_tags.aincrement()
-
-        for digest, man_dc in man_dcs.items():
-            man = await man_dc.resolution()
-            artifact = await sync_to_async(man._artifacts.get)()
-            with artifact.file.open() as content_file:
-                raw = content_file.read()
-            content_data = json.loads(raw)
-            await self.handle_blobs(man_dc, content_data)
-
-        for tag_dc in tag_dcs:
+                    self.tag_dcs.append(tag_dc)
+                    self.manifest_dcs.append(man_dc)
+
+                # Count the skipped tasks as parsed too.
+                await pb_parsed_tags.aincrement()
+
+                # Flush the queues to prevent overly excessive memory usage.
+                # This will cap the number of in flight high level objects to about BATCH_SIZE.
+                if (
+                    len(self.tag_dcs)
+                    + len(self.signature_dcs)
+                    + len(self.manifest_dcs)
+                    + len(self.manifest_list_dcs)
+                    >= BATCH_SIZE
+                ):
+                    await self.resolve_flush()
+
+        await self.resolve_flush()
+
+    async def get_signature_source(self):
+        """
+        Find out where signatures come from: sigstore, extension API or not available at all.
+        """
+        if self.remote.sigstore:
+            return SIGNATURE_SOURCE.SIGSTORE
+
+        registry_v2_url = urljoin(self.remote.url, "v2")
+        extension_check_downloader = self.remote.get_noauth_downloader(url=registry_v2_url)
+        response_headers = {}
+        try:
+            result = await extension_check_downloader.run()
+            response_headers = result.headers
+        except aiohttp.client_exceptions.ClientResponseError as exc:
+            if exc.status == 401:
+                response_headers = dict(exc.headers)
+        if response_headers.get(SIGNATURE_HEADER) == "1":
+            return SIGNATURE_SOURCE.API_EXTENSION
+
+    async def resolve_flush(self):
+        """Resolve pending contents dependencies and put in the pipeline."""
+        # Order matters! Things depended on must be resolved first.
+        for manifest_dc in self.manifest_dcs:
+            config_blob_dc = manifest_dc.extra_data.get("config_blob_dc")
+            if config_blob_dc:
+                manifest_dc.content.config_blob = await config_blob_dc.resolution()
+            for blob_dc in manifest_dc.extra_data["blob_dcs"]:
+                # Just await here. They will be associated in the post_save hook.
+                await blob_dc.resolution()
+            await self.put(manifest_dc)
+        self.manifest_dcs.clear()
```

Comment:
Can we end up in a state where `self.manifest_dcs.clear()` will be invoked less frequently than the actual iteration (because of `await blob_dc.resolution()`)? In the end, we will be iterating over the same manifest declarative contents.

Reply:
I do not understand the question. We collect the manifest_dc in a bucket, we make sure they are ready to processed, we shove them in the pipeline we empty the bucket and start over.
There is no break, so the only thing that can end the loop prematurely is a fatal exception.

Reply:
I meant that if there is any possibility to run `resolve_flush` for the same declarative content objects. So, we will be calling `resolve_flush` on declarative content that was already queued.

Reply:
It can happen that multiple declarative content objects result in the same content, but every dc instance will only be part of a `resolve_flush` call once. (That is why we clear the list.)